### PR TITLE
Read with upper limit based on utc time

### PR DESCRIPTION
**Pull request by adam-yorwerth**

### CODE REVIEW

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 1]
```diff
@@ -500,6 +494,52 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
         noExceptionThrown()
     }
 
+    @Unroll
+    def "when messages have out of order created_utc, we read up to the message with minimum offset outside the limit"() {
```
Comment (KrzysztofKowalczyk) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:498]: That's a nice name

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 2]
```diff
@@ -500,6 +494,52 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
...
+        given:
+        storage = new PostgresqlStorage(dataSource, limit, retryAfter, batchSize, 0, 1, 1)
+        storage.currentTimestamp = "TO_TIMESTAMP( '2000-12-01 10:00:01', 'YYYY-MM-DD HH:MI:SS' )"
+
+        insert(message(1, "type1", "A", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"))
+        insert(message(2, "type1", "B", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"))
+        insert(message(3, "type1", "C", "content-type", ZonedDateTime.parse("2000-12-01T10:00:00Z"), "data"))
+        insert(message(4, "type1", "D", "content-type", ZonedDateTime.parse("2000-12-01T10:00:01Z"), "data"))
```
Comment (KrzysztofKowalczyk) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:506]: We could make it slightly more explicit to point where is the important row, either introduce local variables with some nice names or maybe a comment
```
// <-- out of order
```

File: pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy [hunk 3]
```diff
@@ -500,6 +494,52 @@ class PostgresqlStorageIntegrationSpec extends StorageSpec {
...
+        messageResults.messages*.key == ["A", "B", "C"]
+        messageResults.messages*.offset*.intValue() == [1, 2, 3]
+        messageResults.globalLatestOffset == OptionalLong.of(3)
+
+        where:
+        types << [ [], ["type1"] ]
+    }
+
+    @Unroll
+    def "read up to the last message"(){
```
Comment (KrzysztofKowalczyk) [inline pipe-storage-postgresql/src/integration/groovy/com/tesco/aqueduct/pipe/storage/PostgresqlStorageIntegrationSpec.groovy:526]: that name is not so nice, what is specific about this particular test case ?

File: pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java [hunk 4]
```diff
@@ -295,8 +295,14 @@ private String withInnerJoinToClusters() {
...
-    private static String getSelectLatestOffsetQuery() {
-        return " SELECT coalesce(max(msg_offset),0) as last_offset FROM events WHERE created_utc < CURRENT_TIMESTAMP - ?;";
+    private String getSelectLatestOffsetQuery() {
+        String filterCondition = "WHERE created_utc >= %s - INTERVAL '%s SECONDS'";
+        return
+            "SELECT coalesce (" +
+                " (SELECT min(msg_offset) - 1 FROM events " + String.format(filterCondition, currentTimestamp, readDelaySeconds) + "), " +
+                " (SELECT max(msg_offset) FROM events), " +
+                " 0 " +
+            ") as last_offset;";
```
Comment (KrzysztofKowalczyk) [inline pipe-storage-postgresql/src/main/java/com/tesco/aqueduct/pipe/storage/PostgresqlStorage.java:305]: Gotta love PG planner...

**Review by KrzysztofKowalczyk — APPROVED** [review]
The second test case look like something we had already, might need better name

**Review by adam-yorwerth — COMMENTED** [review]
Looks good to me

**Review by sebastianlavender — APPROVED** [review]

**Pull request merged**